Add elliptic ImposeBoundaryConditions action

nilsvu opened the pull request:
## Proposed changes

This action always imposes zero dirichlet boundary conditions on the linear solver operand, since the non-zero terms are moved to the r.h.s. of the equation as contributions to the source.

The code is mostly copied from `NumericalAlgorithms/DiscontinuousGalerkin/Actions/ImposeBoundaryConditions.hpp` and adapted slightly. It is a rather temporary solution to get boundary conditions working for elliptic systems and will need a rework when more involved boundary conditions will become necessary.

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [x] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [x] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [x] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

nilsvu committed:
Add elliptic ImposeHomogeneousDirichletBoundaryConditions action

mar-celine commented on tests/Unit/Elliptic/DiscontinuousGalerkin/Test_ImposeBoundaryConditions.cpp:
```diff
@@ -0,0 +1,487 @@
+  const CoordinateMaps::Affine xi_map{-1., 1., 3., 7.};
+  const CoordinateMaps::Affine eta_map{-1., 1., 7., 3.};
+
+  const auto coordmap =
```
coord_map

mar-celine commented on tests/Unit/Elliptic/DiscontinuousGalerkin/Test_ImposeBoundaryConditions.cpp:
```diff
@@ -0,0 +1,487 @@
+  //        |       +-+-+   |
+  //        |       | | |   |
+  //        +-------+-+-+---+
+  // We run the actions on the indicated element.
```
"on the element indicated on the diagram." Now, I have seen this diagram in the code before, but I remember others and myself having trouble deciphering it, so I would suggest adding a bit more description on how to read the diagram.

Currently I don't remember how to read the diagram myself :stuck_out_tongue: but I will make a guess:
There are two elements - the element on the left (element 0) has its xi axis pointing upward and eta axis pointing to the left, while the element on the right (element 1) has its xi axis pointing to the right and eta axis pointing downwards.  I remember the first time I deciphered this diagram that it took me a while to even get that far.

Is the indicated element the element indicated by the `x`? I would use the wording "`self_id` is the ElementId (given by the Block ID and segment IDs) of the element indicated by the `x`. After that the rest becomes clear :smile:

nilsvu replied:
There are two blocks (not two elements), xi is pointing right and eta is pointing down. I made a new diagram which is simpler and hopefully clearer.

mar-celine commented on tests/Unit/Elliptic/DiscontinuousGalerkin/Test_ImposeBoundaryConditions.cpp:
```diff
@@ -0,0 +1,487 @@
+                             {{Direction<2>::lower_xi(), {{west_id}, {}}},
+                              {Direction<2>::upper_eta(), {{south_id}, {}}}});
+
+    auto map = ElementMap<2, Frame::Inertial>(self_id, coordmap->get_clone());
```
I think it is a good recommendation to not use names which fit entirely within other names (for example `map` will fit into `coord_map`). It is a good practice to use names not contained in other names so that we can both:
1 - know what distinguishes this variable from the other variable
2 - have an easier time using find and replace if it ever becomes necessary :octocat:

mar-celine commented on tests/Unit/Elliptic/DiscontinuousGalerkin/Test_ImposeBoundaryConditions.cpp:
```diff
@@ -0,0 +1,487 @@
+      get<field_tag>(vars) = data.exterior_bdry_vars.at(direction);
+    }
+
+    int initial_temporal_id = 0;
```
can these be const? should they be `size_t`s?

nilsvu replied:
I set up a temporal id with type `int` at the beginning of the file. It's only for testing, could be any type.

mar-celine commented on tests/Unit/Elliptic/DiscontinuousGalerkin/Test_ImposeBoundaryConditions.cpp:
```diff
@@ -0,0 +1,487 @@
+      Scalar<DataVector>({{{4., 5., 6.}}});
+  CHECK(exterior_vars == expected_vars);
+
+  // ReceiveDataForFluxes
```
Is this comment indicating that the next action is ReceiveDataForFluxes?

nilsvu replied:
yes

mar-celine commented on tests/Unit/Elliptic/DiscontinuousGalerkin/Test_ImposeBoundaryConditions.cpp:
```diff
@@ -0,0 +1,487 @@
+      const std::pair<Direction<2>, ElementId<2>>& mortar_id,
+      const Scalar<DataVector>& local_flux,
+      const Scalar<DataVector>& remote_flux,
+      const Scalar<DataVector>& local_other,
```
What does the `other` refer to? some other non-flux/non-normal quantity? What is it used for?

nilsvu replied:
there is an `OtherData` tag which the test numerical flux defined at the beginning of this file packages into `ExtraData`

nilsvu commented:
@gsb76 I addressed your comments in the fixup commit. Note that most of the testing code is copied from `NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_ImposeBoundaryConditions.cpp`. It's nevertheless good to modernise it. In particular the domain decomposition should be easier to understand now.

mar-celine commented:
@nilsleiffischer Thank you for the changes :+1: In particular the diagram is much clearer now, thanks!

nilsvu commented:
@nilsdeppe @wthrowe @kidder this should be ready then

nilsdeppe commented:
Just to set expectations correctly, given the current backlog of PRs it is unlikely any of us three will have time to review this in the next few weeks, maybe even month or two. However, you should feel free to solicit feedback from other people.

kidder commented:
I should have time to look at this later this week.

wthrowe commented on src/Elliptic/DiscontinuousGalerkin/ImposeBoundaryConditions.hpp:
```diff
@@ -0,0 +1,180 @@
+    constexpr size_t volume_dim = system::volume_dim;
+
+    // Set the data on exterior (ghost) faces to impose the boundary conditions
+    db::mutate_apply<
```
We've generally been using `db::mutate` for inline lambdas and `db::mutate_apply` for externally defined function structs.  I will admit that the two functions are kind of redundant.

nilsvu replied:
I copied this from `NumericalAlgorithms/DiscontinuousGalerkin/Actions/ImposeBoundaryConditions.hpp`. Want me to change it?

nilsdeppe replied:
yea, the majority of the code follows the convention @wthrowe mentioned

wthrowe commented on src/Elliptic/DiscontinuousGalerkin/ImposeBoundaryConditions.hpp:
```diff
@@ -0,0 +1,180 @@
+
+            // By default, use the variables on the external boundary for the
+            // exterior
+            exterior_vars.assign_subset(interior_vars.at(direction));
```
Can this just be `exterior_vars = interior_vars.at(direction);`?

wthrowe commented on src/Elliptic/DiscontinuousGalerkin/ImposeBoundaryConditions.hpp:
```diff
@@ -0,0 +1,180 @@
+              using dirichlet_operand_tag =
+                  LinearSolver::Tags::Operand<dirichlet_tag>;
+              // Use mirror principle. This only works for scalars right now.
+              get<dirichlet_operand_tag>(exterior_vars).get() =
```
Use free function `get`.

wthrowe commented on tests/Unit/Elliptic/DiscontinuousGalerkin/Test_ImposeBoundaryConditions.cpp:
```diff
@@ -0,0 +1,491 @@
+  ActionTesting::MockRuntimeSystem<Metavariables> runner{
+      {NumericalFlux{}}, std::move(dist_objects)};
+
+  using initial_databox_type = db::compute_databox_type<tmpl::append<
```
Probably clearer and easier to maintain if this is written as `decltype(start_box)`.

nilsvu commented:
fixed @wthrowe

nilsvu commented:
done @nilsdeppe @wthrowe

wthrowe commented:
Looks good to me.

nilsdeppe reviewed (commented):
:+1: I did not check this super closely but am fine with it. I expect this to be revisited once the evolution boundary conditions are generalized (I'm working on that next)

nilsvu commented:
That’s the plan @nilsdeppe :)

kidder commented:
go ahead and squash

nilsvu commented:
done @nilsdeppe @wthrowe @kidder 

**Edit:** The one Travis build segfaults repeatedly in tests that are unrelated to this PR.

nilsdeppe commented:
@kidder @gsb76 this is ready for you to give your final feedback :)
